File: histogram.m
clear all
close all
format long

K = 0.5
temperature = 0.1
nbins = 500

D = @(theta) sqrt(K) .* cos(theta).^2 + sin(theta).^2 ./ sqrt(K);
U = @(theta) cos(theta).^2 .* sin(theta).^2 / K;
%U = @(theta) sin(theta);
f = @(theta) sqrt(D(theta)) .* exp(-U(theta) ./ temperature);

N = 1e7;

z1 = quadgk(f, -pi, pi)

% ts = linspace(-pi, pi, N*10);
% h = ts(2) - ts(1);
% z2 = h*trapz(f(ts))

% zerr = log10(abs(z1-z2))

hst1 = zeros(1, nbins);
% hst2 = zeros(1, nbins);
intervals = linspace(-pi, pi, nbins + 1);

for r = 1:nbins
    % disp(['Step ' num2str(r)]);
    
    a = intervals(r);
    b = intervals(r+1);
    hst1(r) = quadgk(f, a, b) / z1;

    % ts = linspace(a, b, N);
    % dt = ts(2) - ts(1);
    % hst2(r) = dt * trapz(f(ts)) / z2;
end

% err = log10(norm(hst1 - hst2, 'inf'))

% subplot(1, 2, 1);
plot(intervals(2:end), hst1);
data = [intervals(1:end-1)' hst1'];
save('histogram.dat', '-ascii', 'data')
% subplot(1, 2, 2);
% hist(hst2);

% hst3 = [0.06335038037960476 0.07846018979248426 0.08547751219967743 0.0803223327528606 0.06946234526594158 0.06946234526594158 0.08032233275286062 0.08547751219967741 0.07846018979248426 0.06335038037960475 0.04648530341829146 0.03194415120522183 0.02106348636216899 0.01370652625218747 0.009727772371561617 0.009727772371561617 0.01370652625218747 0.02106348636216898 0.03194415120522182 0.04648530341829145];
% %hst3 = [0.0676574 0.0650374 0.0551709 0.0418376 0.031793 0.031793 0.0418376 0.0551709 0.0650374 0.0676574 0.0637844 0.0565438 0.0479972 0.0387983 0.0313801 0.0313801 0.0387983 0.0479972 0.0565438 0.0637844];
% err13 = log10(norm(hst1 - hst3, 'inf'))
% err23 = log10(norm(hst2 - hst3, 'inf'))

File: Histogram.cpp
#include <stdio.h>
#include <math.h>
#include <gsl/gsl_integration.h>

#include <vector>
#include <sstream>
#include <iostream>
#include <iomanip>
#include <iterator>
#include <numeric>
#include <algorithm>
#include <functional>

#include "linspace.h"
#include "Plotter.h"
#include "Histogram.h"

using std::vector;

static double f(double theta, void* params) {
  const double temperature = *((double *) params);
  const double K = *((double *) params + 1);

  double s, c;
  sincos(theta, &s, &c);
  const double D = sqrt(K) * c * c + s * s / sqrt(K);
  const double U = s;
  
  return sqrt(D) * exp(-U / temperature);
}
  
void Histogram::compute() {
  double params[] = { temperature, K };
  
  gsl_function F;
  F.function = &f;
  F.params = &params[0];

  double err;
  double partition_function;
  vector<double> intervals = linspace<double>(-M_PI, M_PI, nbins+1);
  
  gsl_integration_qags(&F, -M_PI, M_PI, 0, 1e-12, 10000,
                       w, &partition_function, &err);
  
  for (size_t n = 0; n < histogram.size(); n++) {
    const double a = intervals[n], b = intervals[n+1];

    gsl_integration_qags(&F, a, b, 0, 1e-12, 10000,
                         w, &reference[n], &err);

    reference[n] /= partition_function;
  }
  
  minimum = *std::min_element(reference.begin(), reference.end());
  maximum = *std::max_element(reference.begin(), reference.end());

#if 0
  std::cout << std::setprecision(16)
            << "Partition function = "
            << partition_function << "\n";
  for (size_t k = 0; k < nbins; k++)
    std::cout << reference[k] << " ";
  std::cout << endl;
#endif
}

unsigned long& Histogram::operator[](double theta) {
  const double a = -M_PI, b = M_PI;
  const double N = static_cast<double>(nbins);
  const double i = floor((theta - a) / (b - a) * N);

  return histogram.at(static_cast<size_t>(i));
}

unsigned long long Histogram::total() const {
  // return std::accumulate(histogram.begin(), histogram.end(),
  //                        0, std::plus<unsigned long>());
  unsigned long long count = 0;

  for (size_t k = 0; k < histogram.size(); k++)
    count += histogram[k];

  return count;
}

double Histogram::error() const {
  // A composition of std::transform and std::max_element might be
  // more elegant here.
  const double tot = static_cast<double>(total());
  double err = -INFINITY;               // Error in infty-norm.
  for (size_t k = 0; k < nbins; k++) {
    const double freq = static_cast<double>(histogram[k]) / tot;
    err = std::max(err, fabs(freq - reference[k]));
  }
  return err;
  
#if 0
  double err = 0.0;
  for (size_t k = 0; k < nbins; k++)
    err += fabs(histogram[k] / tot - reference[k]);
  return err / static_cast<double>(nbins);
#endif
}

std::ostream& operator<<(std::ostream& stream, const Histogram& h) {
  const auto tot = h.total();

  stream << std::setprecision(16) << tot << " ";
  for (size_t k = 0; k < h.nbins; k++)
    stream << h.histogram[k] << " ";

  return stream << std::endl;
}

static inline double angle(size_t k, size_t nbins) {
  return 2.0 * M_PI / (double) nbins * (double) k - M_PI;
}

void Histogram::plot(Plotter& plotter) const {
  const double tot = static_cast<double>(total());
  const double order_error = log10(error());

  std::ostringstream cmd;
  cmd << "set title 'O(Error) = " << order_error << "'\n"
      << "set xrange [-pi:pi]\n"
      << "set yrange [" << 0.0 << ":" << (maximum + minimum) << "]\n"
      << "plot '-' with linespoints pointtype 5"
      << " linecolor rgbcolor 'black' title 'Reference',"
      << " '-' with linespoints pointtype 7"
      << " linecolor rgbcolor 'blue' title 'Simulation'\n";
  for (unsigned k = 0; k < nbins; k++)
    cmd << angle(k, nbins) << " " << reference[k] << "\n";
  cmd << "\ne\n";
  for (unsigned k = 0; k < nbins; k++)
    cmd << angle(k, nbins) << " "
        << static_cast<double>(histogram[k]) / tot << "\n";
  cmd << "\ne\n";
  plotter.send(cmd.str());
}
